### ISSUE TITLE
Improper handling of subprocess stream causes Gradle to hang

### ISSUE THREAD

**Issue by blindpirate** [gradle/gradle#6114]
### Context

Recently we've investigated the issue of build execution hanging. The stack trace is:

```
 "gradle-shutdown-hook" #18 prio=5 os_prio=0 tid=0x0000000055320800 nid=0x1b24 waiting on condition [0x000000005521e000]
 java.lang.Thread.State: WAITING (parking)
 	at sun.misc.Unsafe.park(Native Method)
 	- parking to wait for  <0x00000000feef70a0> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
 	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
 	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
 	at org.gradle.process.internal.DefaultExecHandle.waitForFinish(DefaultExecHandle.java:300)
 	at org.gradle.process.internal.DefaultExecHandle.abort(DefaultExecHandle.java:289)
 	at org.gradle.process.internal.ExecHandleShutdownHookAction.run(ExecHandleShutdownHookAction.java:39)
 	at org.gradle.process.internal.shutdown.ShutdownHookActionRegister$GradleShutdownHook.run(ShutdownHookActionRegister.java:41)
 	at java.lang.Thread.run(Thread.java:748)
 
 "pool-38-thread-3" #337 prio=5 os_prio=0 tid=0x0000000054f7f800 nid=0x1c14 runnable [0x000000006116e000]
 java.lang.Thread.State: RUNNABLE
 	at java.io.FileInputStream.readBytes(Native Method)
 	at java.io.FileInputStream.read(FileInputStream.java:233)
 	at org.gradle.process.internal.streams.ExecOutputHandleRunner.forwardContent(ExecOutputHandleRunner.java:61)
 	at org.gradle.process.internal.streams.ExecOutputHandleRunner.run(ExecOutputHandleRunner.java:51)
 	at org.gradle.internal.operations.CurrentBuildOperationPreservingRunnable.run(CurrentBuildOperationPreservingRunnable.java:38)
 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
 	at java.lang.Thread.run(Thread.java:748)
 
 "pool-38-thread-2" #336 prio=5 os_prio=0 tid=0x0000000054f79800 nid=0x1ecc runnable [0x0000000060abf000]
 java.lang.Thread.State: RUNNABLE
 	at java.io.FileInputStream.readBytes(Native Method)
 	at java.io.FileInputStream.read(FileInputStream.java:255)
 	at java.io.BufferedInputStream.fill(BufferedInputStream.java:246)
 	at java.io.BufferedInputStream.read1(BufferedInputStream.java:286)
 	at java.io.BufferedInputStream.read(BufferedInputStream.java:345)
 	- locked <0x00000000feef28c0> (a java.io.BufferedInputStream)
 	at java.io.FilterInputStream.read(FilterInputStream.java:107)
 	at org.gradle.process.internal.streams.ExecOutputHandleRunner.forwardContent(ExecOutputHandleRunner.java:61)
 	at org.gradle.process.internal.streams.ExecOutputHandleRunner.run(ExecOutputHandleRunner.java:51)
 	at org.gradle.internal.operations.CurrentBuildOperationPreservingRunnable.run(CurrentBuildOperationPreservingRunnable.java:38)
 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
 	at java.lang.Thread.run(Thread.java:748)
 
 "pool-38-thread-1" #334 prio=5 os_prio=0 tid=0x0000000055316800 nid=0x2c24 waiting on condition [0x000000006093e000]
 java.lang.Thread.State: WAITING (parking)
 	at sun.misc.Unsafe.park(Native Method)
 	- parking to wait for  <0x00000000feef1db8> (a java.util.concurrent.CountDownLatch$Sync)
 	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
 	at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
 	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:997)
 	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
 	at java.util.concurrent.CountDownLatch.await(CountDownLatch.java:231)
 	at org.gradle.process.internal.streams.OutputStreamsForwarder.stop(OutputStreamsForwarder.java:68)
 	at org.gradle.process.internal.DefaultExecHandle$CompositeStreamsHandler.stop(DefaultExecHandle.java:429)
 	at org.gradle.process.internal.ExecHandleRunner.run(ExecHandleRunner.java:81)
 	at org.gradle.internal.operations.CurrentBuildOperationPreservingRunnable.run(CurrentBuildOperationPreservingRunnable.java:38)
 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
 	at java.lang.Thread.run(Thread.java:748)
 
 
 "main" #1 prio=5 os_prio=0 tid=0x0000000000a3f000 nid=0xe34 in Object.wait() [0x0000000000c1e000]
 java.lang.Thread.State: WAITING (on object monitor)
 	at java.lang.Object.wait(Native Method)
 	- waiting on <0x00000000e07b4ee8> (a java.lang.Thread)
 	at java.lang.Thread.join(Thread.java:1252)
 	- locked <0x00000000e07b4ee8> (a java.lang.Thread)
 	at java.lang.Thread.join(Thread.java:1326)
 	at java.lang.ApplicationShutdownHooks.runHooks(ApplicationShutdownHooks.java:106)
 	at java.lang.ApplicationShutdownHooks$1.run(ApplicationShutdownHooks.java:46)
 	at java.lang.Shutdown.runHooks(Shutdown.java:123)
 	at java.lang.Shutdown.sequence(Shutdown.java:167)
 	at java.lang.Shutdown.exit(Shutdown.java:212)
 	- locked <0x00000000e0745ff0> (a java.lang.Class for java.lang.Shutdown)
 	at java.lang.Runtime.exit(Runtime.java:109)
 	at java.lang.System.exit(System.java:971)
 	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:68)
```

Obviously, the process hangs at [this line](https://github.com/gradle/gradle/blob/2bafa9609490a29020eeeee4fe0c9a176afd8adc/subprojects/core/src/main/java/org/gradle/process/internal/ExecHandleRunner.java#L81).

```
int exitValue = process.waitFor();
streamsHandler.stop(); <--- stuck at this line
```

which means, although subprocess already returns (`waitFor()` returns), its `stdout` and `stderr` streams are still in block state. This is rare but not impossible. Searched StackOverflow we found two similar issues: 

- https://stackoverflow.com/questions/9012639/java-process-inputstream-stuck
- https://stackoverflow.com/questions/9310077/main-program-hang-when-read-output-stream-of-processb-which-starts-another-proce

On this scenario, subprocess' stdout and stderr stream should not be read any more because the subprocess no longer exists.

### Steps to Reproduce (for bugs)

To reproduce this bug, you have to have JDK6 installed and be on Windows.

```
import org.gradle.process.internal.ExecHandleFactory
import org.gradle.internal.classloader.ClasspathUtil

class ParentProcess {
   static String JAVA6 = "C:\\Program Files\\Java\\jdk1.6.0_45\\bin\\java.exe"
   public static void main(String[] args) throws Exception {
       System.out.println("Before call child process");
       ProcessBuilder builder = new ProcessBuilder(JAVA6, '-cp', System.getProperty('java.class.path'), ChildProcess.class.name);
       builder.start();
       System.out.println("After call child process");
   }
}

public class ChildProcess {
   public static void main(String[] args) throws Exception {
       Thread.sleep(3600*1000);
   }
}

task hang {
   doLast {
       String classpath = ClasspathUtil.getClasspath(ParentProcess.class.classLoader)
               .asFiles
               .collect { it.absolutePath}
               .join(';')

       def execHandler = project.services.get(ExecHandleFactory).newExec()
               .executable(ParentProcess.JAVA6)
               .args('-cp', classpath, ParentProcess.class.name)
               .build()

       execHandler.start()

       Thread.sleep(5000)
       execHandler.abort()
   }
}
```

### Your Environment

Currently this issue is only observed on Windows.

**Comment by big-guy** [comment]
From your stacktrace, it seems like this is the scenario?

`Gradle Daemon` -> `Gradle worker` -> `some other process` -> `child process`

And we're seeing the hang in `Gradle worker` because `child process` isn't stopped completely?

If this is something we think is fixed in Java 7 (WRT `inheritIO()` and `ProcessBuilder`), maybe this is something we can try to fix when we bump the minimum version of Java for Gradle?

**Comment by blindpirate** [comment]
Thanks @big-guy .

> Gradle Daemon -> Gradle worker -> some other process -> child process

Exactly.

> And we're seeing the hang in Gradle worker because child process isn't stopped completely?

I'm not sure if "child process isn't stopped completely", I think given that `waitFor()` returns, the child process is stopped completely, right?

Currently I can only reproduce it on Java 6, I'm wondering, does this only happen on Java (below 6)? Can it happen on any other non-java process? I searched our codebase and found out some usage of Java6, for example, https://github.com/gradle/gradle/blob/80893b17410bf865e1c8cb813cc87592a915eff0/subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/ToolingApiUnsupportedClientJvmCrossVersionSpec.groovy#L76 . If this is caused by Java6, I think it acceptable to disable such tests on Windows.

**Comment by big-guy** [comment]
Windows is strange in that it doesn't really have parent/child processes in the same way that Linux does.  I know we've had in the past problems with child processes outliving the parents because the parents could be killed independently of the children.

Are the hanging builds like this restricted to just the `tooling-api` project?  I wouldn't expect that test to have the problem that you're seeing (it shouldn't be starting any separate workers).  Do we have any idea what the child process from the worker was? 

This looks related (and fixed in Java 7+): https://bugs.openjdk.java.net/browse/JDK-7147084

**Comment by blindpirate** [comment]
Thanks @big-guy for the information. 

> Are the hanging builds like this restricted to just the tooling-api project?

https://builds.gradle.org/viewLog.html?buildId=14531791&buildTypeId=Gradle_Check_Platform_Java8_Oracle_Windows_workers&tab=buildResultsDiv . `ps` also says it was running `workers` subproject.

```
 "C:\Program Files\Java\jdk1.8\bin\java.exe" "-DintegTest.gradleHomeDir=C:\tcagent1\work\668602365d1521fc\subprojects\workers\build\integ test" -DintegTest.gradleUserHomeDir=C:\tcagent1\work\668602365d1521fc\intTestHomeDir -DintegTest.toolingApiShadedJarDir=C:\tcagent1\work\668602365d1521fc\subprojects\tooling-api\build\shaded-jar -Djava.security.manager=worker.org.gradle.process.internal.worker.child.BootstrapSecurityManager "-Djava9Home=C:\Program Files\Java\jdk1.9" -Dorg.gradle.ci.agentCount=2 -Dorg.gradle.ci.agentNum=1 -Dorg.gradle.integtest.daemon.registry=C:\tcagent1\work\668602365d1521fc\build\daemon -Dorg.gradle.integtest.executer=forking -Dorg.gradle.integtest.mirrors.google=http://dev12.gradle.org:8081/artifactory/google -Dorg.gradle.integtest.mirrors.gradle=http://dev12.gradle.org:8081/artifactory/gradle-repo/ -Dorg.gradle.integtest.mirrors.gradle-releases=http://dev12.gradle.org:8081/artifactory/gradle-releases/ -Dorg.gradle.integtest.mirrors.gradle-snapshots=http://dev12.gradle.org:8081/artifactory/gradle-snapshots/ -Dorg.gradle.integtest.mirrors.gradlejavascript=http://dev12.gradle.org:8081/artifactory/gradle-javascript/ -Dorg.gradle.integtest.mirrors.gradleplugins=http://dev12.gradle.org:8081/artifactory/gradle-plugins/ -Dorg.gradle.integtest.mirrors.jboss=http://dev12.gradle.org:8081/artifactory/jboss/ -Dorg.gradle.integtest.mirrors.jcenter=http://dev12.gradle.org:8081/artifactory/jcenter -Dorg.gradle.integtest.mirrors.lightbendivy=http://dev12.gradle.org:8081/artifactory/typesafe-ivy-releases -Dorg.gradle.integtest.mirrors.lightbendmaven=http://dev12.gradle.org:8081/artifactory/typesafe-maven-releases -Dorg.gradle.integtest.mirrors.mavencentral=http://dev12.gradle.org:8081/artifactory/repo1 -Dorg.gradle.integtest.mirrors.restlet=http://dev12.gradle.org:8081/artifactory/restlet/ -Dorg.gradle.integtest.mirrors.springreleases=http://dev12.gradle.org:8081/artifactory/spring-releases/ -Dorg.gradle.integtest.mirrors.springsnapshots=http://dev12.gradle.org:8081/artifactory/spring-snapshots/ -Dorg.gradle.integtest.mirrors.typesafeivy=http://dev12.gradle.org:8081/artifactory/typesafe-ivy-releases -Dorg.gradle.integtest.mirrors.typesafemaven=http://dev12.gradle.org:8081/artifactory/typesafe-maven-releases -Dorg.gradle.integtest.multiversion=all -Dorg.gradle.integtest.native.toolChains=all -Dorg.gradle.integtest.versions=latest -Dorg.gradle.native=false -Dorg.gradle.test.maxParallelForks=4 -XX:+HeapDumpOnOutOfMemoryError -Xmx512m -Dfile.encoding=UTF-8 -Duser.country=US -Duser.language=en -Duser.variant -ea -cp C:\Users\tcagent1\.gradle\caches\4.10-20180725080856+0000\workerMain\gradle-worker.jar worker.org.gradle.process.internal.worker.GradleWorkerMain "'Gradle Test Executor 393'"       9824 
```

I didn't find any jdk6 usage in `workers`, which makes me quite puzzled. 

> Do we have any idea what the child process from the worker was?

Since the child process already exits (`waitFor` returns), I wouldn't expect to see it in the `ps` result, but every process left on that machine looks pretty normal (and one of them is supposed to be the grandchild process): 

- 2 TeamCity JVMs
- 3 Gradle daemons started by test workers
- 1 Gradle worker daemon
- 1 Gradle wrapper
- 1 Gradle daemon which runs the build
- 1 JVM which runs `jstack`

Is there anything we can do on Gradle side?

The attachment is the ps and jsta
[Test_Coverage_-_Platform_Java8_Oracle_Windows_Test_Coverage_-_Platform_Java8_Oracle_Windows_workers_9023.log](https://github.com/gradle/gradle/files/2244142/Test_Coverage_-_Platform_Java8_Oracle_Windows_Test_Coverage_-_Platform_Java8_Oracle_Windows_workers_9023.log)

**Comment by blindpirate** [comment]
@big-guy Seems like https://github.com/gradle/gradle/blob/c8143c9bcf412a2d19053b8c926c7787c1dc8087/subprojects/core/src/main/java/org/gradle/util/DisconnectableInputStream.java#L31 is what we need here, right?

**Comment by blindpirate** [comment]
BTW, I tested with `DisconnectableInputStream` on my windows machine and looks like it works.

**Comment by blindpirate** [comment]
https://builds.gradle.org/viewLog.html?buildId=15082559&buildTypeId=Gradle_Check_Platform_Java8_Oracle_Windows_workers&tab=buildResultsDiv

**Comment by blindpirate** [comment]
Superseded by https://github.com/gradle/gradle-private/issues/1486
